### ISSUE TITLE
No issue - just a Heads-Up - New IQFeed limit on history requests

### ISSUE THREAD

**Issue by BunkerCoder** [mathpaquette/IQFeed.CSharpApiClient#90]
Today, I received "E,Too many simultaneous history requests" in response to an HTD request.
I called support and they said it is a recent change, and they will be posting it on the developer forum soon.
The change was explained to me as:
Basically each user has a limit of 30 history requests per second, and every 1/30 of a second an existing request "drops off" and it decreases by one. As long you stay below an average of 30 request/second (and there is a little padding) you'll be fine.

Just wanted to pass it along.

**Comment by mathpaquette** [comment]
Hi BunkerCoder,

I noticed that also this weekend while doing download for the previous week. Raised support dev issue but didn’t receive update Yet. Thanks for sharing with us... looks like we will need something to prevent that...

Thanks

Sent from my iPhone

On Dec 21, 2020, at 5:41 PM, BunkerCoder <notifications@github.com> wrote:

﻿

Today, I received "E,Too many simultaneous history requests" in response to an HTD request.
I called support and they said it is a recent change, and they will be posting it on the developer forum soon.
The change was explained to me as:
Basically each user has a limit of 30 history requests per second, and every 1/30 of a second an existing request "drops off" and it decreases by one. As long you stay below an average of 30 request/second (and there is a little padding) you'll be fine.

Just wanted to pass it along.

**Comment by mathpaquette** [comment]
do you think this will be also required for ticks. Ill be adding required changes to avoid that.

**Comment by BunkerCoder** [comment]
@mathpaquette Thanks for taking this on.  Got this from support: By the way this is only history requests. Port 5009 (Live) will not be changed.

**Comment by BunkerCoder** [comment]
Also per support: It's on a "user + IP address" basis
So the limit is across all clients.

**Comment by mathpaquette** [comment]
@BunkerCoder created a fix. Not tested yet. Do you have some time to test it out?

https://github.com/mathpaquette/IQFeed.CSharpApiClient/commit/a33eb8305589c7179c283fedbeacbdb5b6ad8c4b

**Comment by mathpaquette** [comment]
@BunkerCoder have a look https://github.com/mathpaquette/IQFeed.CSharpApiClient/commit/21cd24b2294fb5ce89c40d5b7b66e807731501c0

**Comment by BunkerCoder** [comment]
@mathpaquette I'm bogged down today on a client fire, and everyone is on vacation!  I'll try to look later today. Thanks for jumping on this.  I will too as soon as I can.

**Comment by BunkerCoder** [comment]
@mathpaquette I ran "git checkout 21cd24b", then built, then re-targeted my throwing app to use the local library.  Throwing app still throws with "E,Too many simultaneous history requests."  Thrown from BaseLookupFacade.cs:line 61 as expected.  Will do some more digging.

**Comment by mathpaquette** [comment]
> @mathpaquette I ran "git checkout [21cd24b](https://github.com/mathpaquette/IQFeed.CSharpApiClient/commit/21cd24b2294fb5ce89c40d5b7b66e807731501c0)", then built, then re-targeted my throwing app to use the local library. Throwing app still throws with "E,Too many simultaneous history requests." Thrown from BaseLookupFacade.cs:line 61 as expected. Will do some more digging.

https://github.com/mathpaquette/IQFeed.CSharpApiClient/commit/21cd24b2294fb5ce89c40d5b7b66e807731501c0#r45348806

**Comment by BunkerCoder** [comment]
Tried with 10. Same result.

**Comment by mathpaquette** [comment]
@BunkerCoder are you able to reproduce now? Im not able to get the exception.... Can you share a sample with me to reproduce it?

**Comment by BunkerCoder** [comment]
@mathpaquette  I've started looking at it, Tried adding a stress test in ConcurrentHistoricalExample.cs, and could not get it to fail, so assuming that your solution is working, and I've got a bug in my app.  Will advise results.

**Comment by mathpaquette** [comment]
> @mathpaquette I've started looking at it, Tried adding a stress test in ConcurrentHistoricalExample.cs, and could not get it to fail, so assuming that your solution is working, and I've got a bug in my app. Will advise results.

no I dont think so. I think they simply turned it off for now. Lets see. Tell me if you notice something on your side... Thanks

**Comment by BunkerCoder** [comment]
Now that I look again, I think you're right. I re-tested my throwing app and it didn't throw.

**Comment by mathpaquette** [comment]
@BunkerCoder I think their implementation was broken and they decided to roll it back.. lets see.

**Comment by BunkerCoder** [comment]
Received DTN email;  rate limit removed until 12-Jan, when 50 requests/sec. limit goes into effect.  No word about a test server.

Maybe they will open up an API call to check the rate; multiple simultaneous apps/clients can't see the overall machine rate.
At a minimum, I think that LookupRateLimiter() should be a shared instance or static instead of per-Client.

**Comment by mathpaquette** [comment]
50 requests/sec isssssssss crazy fast. 3000 reqs per minute... we should be good to go. 
right now, every class instances from the same LookupClient is using the same rate limiter. Do you have use case of having multiple LookupClient per program?

**Comment by BunkerCoder** [comment]
Not yet, but I'm working on one! If it becomes a problem, I'll workup a solution and submit a PR.  Maybe just change 30 to 50 for now?

**Comment by BunkerCoder** [comment]
Strike that. Don't see a good use case for multiple LookupClient per program. But may possibly see running multiple programs, so being able to tweak each program would be nice to have. Maybe a constructor that supported:
var lookupRateLimiter = new LookupRateLimiter(requestsPerSecondLimit);
Happy to contribute.

**Comment by mathpaquette** [comment]
Yes absolutely. It’s gonna be a parameter from the factory for sure!

Thanks

Sent from my iPhone

On Dec 23, 2020, at 7:43 PM, BunkerCoder <notifications@github.com> wrote:

﻿

Strike that. Don't see a good use case for multiple LookupClient per program. But may possibly see running multiple programs, so being able to tweak each program would be nice to have. Maybe a constructor that supported:
var lookupRateLimiter = new LookupRateLimiter(requestsPerSecondLimit);
Happy to contribute.

**Comment by BunkerCoder** [comment]
@mathpaquette Silly of me to have even mentioned it. Have I said Thank You lately?

**Comment by mathpaquette** [comment]
@BunkerCoder your involvement is significantly more meaningful than few words :)

**Comment by Nucs** [comment]
Too bad I missed this,
What a mess this change has created.

**Comment by mathpaquette** [comment]
guys.... spent few hours thinking about of one precise working solution... I think I got something.... Ill sharing today.

**Comment by BunkerCoder** [comment]
Many thanks for a very elegant, ,highly-performant, and timely-provided solution!

**Comment by mathpaquette** [comment]
Apparently, 50 requests per second is too high... I'm getting , ErrorMessage: Too many simultaneous history requests., MessageTrace: E,Too many simultaneous history requests.,
!ENDMSG!,

**Comment by mathpaquette** [comment]
Apparently 50 requests per second works well but you have to limit to 8 concurrent clients.

**Comment by BunkerCoder** [comment]
I got the error with 6 concurrent clients @ 50/sec. a couple of times this week.  Instead of fighting it, I dialed back to 45/sec., and haven't seen the error since.
